### PR TITLE
Adding a new test for a empty key file.

### PR DESCRIPTION
**Pull request by douglasAtJoyent**
Adding a test for giving the client a empty key file.

This test should fail currently because of an issue that the client has.

### CODE REVIEW

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                    + "vtPiYO6ezFC3S42iyzTEqy+yMJV+KwE7oKxlQdoGyqCM80TMxcjeorY2rkG9GWTa\n"
+                    + "qx40Tz9df9w8IEUrZLZqgdzLOTf/O0bzUwkn3UwvSGUrC1CUeEAUcYqeIXd9IzPe\n"
+                    + "5NLxgAC02MTtgddqTS1UKb1pAoGANgYAw/utQywzTRie4CfFQZXj8OM78yte1wV4\n"
+                    + "3/Zc6C2y647NguqEkYchEF75MwEPAGCg1Na6F6i5mU/0aJ5ym8EF21ikxlPnB0rn\n"
+                    + "Cb+kHE7HHs9aoyRhBY9FcTgqDDZAq38kprVPYN+rzGrwVK2S2Dm/tuXP6FkabuD8\n"
+                    + "dr6qJJECgYBKc09htDe/OBPqyVtEuUv0HTA5KRNpqeheWCKTDAFtjpciYa1qLyjs\n"
+                    + "JET3gPfjWkcuDOCs6zcaj/anlr/PxdSxYw4Zu21YKZGY70ceBEJgtX5PcrgmpH7R\n"
+                    + "99+DOAhlNqwK5X+4Foo3tErRveaTczgz38G17IfoKJ7rKYHQt1o+bQ==\n" + "-----END RSA PRIVATE KEY-----";
+            bw.write(validKey);
+        } catch (Exception e) {
```
Comment (tjcelaya) [inline src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: These two lines just add noise and are not needed since the exception is being rethrown without any modification, please avoid this pattern in the future:

```java
} catch (Exception e) {
  throw e;
```

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            throw e;
+        }
+        System.setProperty(MANTA_KEY_PATH_PROPERTY, keyFile.getAbsolutePath());
+        MantaClient client = MANTA_CLIENT_SUPPLIER.get();
+        Stream<MantaObject> mor = client.listObjects("/" + System.getProperty(MANTA_USER) + "/stor");
```
Comment (tjcelaya) [inline src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: Two things here:
- This stops being a unit test the moment it attempts to contact an external system. This kind of test belongs in `src/integration` if we intend to keep it.
- Streams tie up client connections so each call to `MantaClient#listObjects` will diminish the connections available in the client's connection pool. If you look at how `MantaTransferClient` you'll notice the supplier is only ever used to build a single client which is reused. The `MantaClient` class is thread-safe and reusing a single instance is encouraged. Creating a new client for each test case both masks the connection pool leak and skips using the client in the way that it is typically utilized.

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    public void invalidUser() throws IOException {
+        System.setProperty(MANTA_USER_PROPERTY, "invalid");
+        MantaClient client = MANTA_CLIENT_SUPPLIER.get();
+        Stream<MantaObject> mor = client.listObjects(base_dir);
+        Assert.assertTrue(mor.count() > 0, "Seeing more than 0 entries, This should have thrown an exception by now.");
+    }
+
+    @Test(expectedExceptions = MantaClientHttpResponseException.class, expectedExceptionsMessageRegExp = ".*HTTP request failed to:.*")
+    public void invalidPassword() throws IOException {
+        System.setProperty(MANTA_USER_PROPERTY, "invalid");
```
Comment (tjcelaya) [inline src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: This test is about password but this line seems to be setting the username? Changing it to "manta.password" triggers this exception which is probably what you expected to be testing:

```
com.joyent.http.signature.KeyLoadException: Loaded key is not encrypted but a password was supplied.
```

Either way, password are used when loading keys, I have no idea why you would expect a `MantaClientHttpResponseException` for a bad password. You'll also need an encrypted private key to properly test the password setting.

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Test;
+
+import com.joyent.http.signature.KeyLoadException;
+import com.joyent.manta.client.MantaClient;
+import com.joyent.manta.client.MantaObject;
+import com.joyent.manta.exception.ConfigurationException;
+import com.joyent.manta.exception.MantaClientHttpResponseException;
+
+import jline.internal.Log;
```
Comment (tjcelaya) [inline src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: This is an immediate red flag for me, why are we using a class in a package with `internal` in the name? My editor tells me the `private static Logger` declared in the class is never used, we should go over how your editor is configured and make sure it shows warnings for unused variables.

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    public static final String MANTA_UPLOAD_BUFFER_SIZE = "manta.upload_buffer_size";
+    public static final String MANTA_SKIP_DIRECTORY_DEPTH = "manta.skip_directory_depth";
+    public static final String MANTA_CLIENT_ENCRYPTION = "manta.client_encryption";
+    public static final String MANTA_CLIENT_ENCRYPTION_KEY_ID = "manta.encryption_key_id";
+    public static final String MANTA_ENCRYPTION_ALGORITHM = "manta.encryption_algorithm";
+    public static final String MANTA_UNENCRYPTED_DOWNLOADS = "manta.permit_unencrypted_downloads";
+    public static final String MANTA_ENCRYPTION_AUTH_MODE = "manta.encryption_auth_mode";
+    public static final String MANTA_ENCRYPTION_KEY_PATH = "manta.encryption_key_path";
+    public static final String MANTA_ENCRYPTION_KEY_BYTES_64 = "manta.encryption_key_bytes_base";
+    private static final String base_dir = String.format("/%s", System.getenv(MANTA_USER));
```
Comment (tjcelaya) [inline src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: This fails to account for subusers, I think you're looking for `com.joyent.manta.client.MantaUtils#parseAccount`

**Review by tjcelaya — CHANGES_REQUESTED** [review]
Some of these tests could be useful but overall I'm not really sure what to make of the latest additions to this PR. I was under the impression we had discussed adding unit tests that were focused on misusing the key-related configuration options but it appears the scope has changed drastically.

I'd be happy to go over general principles for unit testing. Each test case should have a clear purpose and clearly communicate the system under test.

16 of the 18 tests in this class are failing for a variety of reasons. Should they all be passing, all be failing?

Perhaps we should pair up to define a set of test cases and work through implementing them together?

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+
+
+    @BeforeMethod
+    private void setParameters() {
```
[the remark this answers is not in the document]
Comment (douglasAtJoyent) [reply src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: Yes, it assumes that you have a valid setup, I assumed that would be true since other tests also assume that you have a valid setup and a directory in the user's stor directory created.

[the remark this answers is not in the document]
Comment (tjcelaya) [reply src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: Which other tests are you referencing? Only the tests in the `src/integration` folder depend on a properly-configured environment. None of the other tests in `src/test` require valid credentials. Try running `mvn clean test` in a shell without any of the `MANTA_*` variables set.

[the remark this answers is not in the document]
Comment (douglasAtJoyent) [reply src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: Sorry, I think I just assumed they did since the first time I ran them they failed because a folder was not created in my /stor directory.

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java (hunk 4 above)
Comment (douglasAtJoyent) [reply src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: I had removed the logging lines, I was using it earlier and forgot to remove this.

Comment (douglasAtJoyent) [reply src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: Actually I take that back, Line 168, Log.info(it.next());  But I will remove it. Plus I think the import is probably incorrect, that I think I just let my IDE import.

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (tjcelaya) [reply src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: Please be more specific, are you talking about TransferManagerIT? I get this exception when I run the test so I'm assuming that's the one:

```
com.joyent.manta.archiver.TransferClientException: Remote path does not exist: /tomas.celaya/stor/manta-archiver-test/
```

Unfortunately this seems to be a problem with how that test was designed, it should actually be creating the directory during test setup in the same fashion as the integration tests in java-manta.

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+import com.joyent.manta.exception.MantaClientHttpResponseException;
+
+import jline.internal.Log;
+
+@Test
+public class ManataClientSupplierTest {
+    private static Logger LOG = LoggerFactory.getLogger(ManataClientSupplierTest.class);
+
+    private static final MantaClientSupplier MANTA_CLIENT_SUPPLIER = new MantaClientSupplier();
+    // Seems silly, but what if they change?
```
[the remark this answers is not in the document]
Comment (douglasAtJoyent) [reply src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: Agreed, I will change them, I put them there more to have a easy reference.

File: src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (douglasAtJoyent) [reply src/test/java/com/joyent/manta/archiver/ManataClientSupplierTest.java]: How would I create a client without having the environment setup? If I try to create an instance of the client with no environmental variables set I get an exception.

**Review by tjcelaya — APPROVED** [review]

**Comment by tjcelaya** [comment]
Sorry this took so long to get to. I pushed a commit with some small formatting fixes to make the annotation arguments more legible and switched from `KeyLoadException` to `ConfigurationException` (since the `KeyLoadException` is caught and wrapped by `MantaClient`).

Thank you for contributing!

**Comment by tjcelaya** [comment]
(I'll leave the branch for you to delete if you want. You can reuse it in the future but it can lead to confusion if you don't merge master back into it.)
